sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,156 @@
+
+        private TValue _value;
+
+        private RequestFailedException _requestFailedException;
+
+        public OperationInternal(string operationTypeName,
+            ClientDiagnostics clientDiagnostics,
+            Func<CancellationToken, Task<Response<TResult>>> getResponseAsync,
+            Func<CancellationToken, Response<TResult>> getResponse,
+            Func<Response<TResult>, OperationInternalResponseStatus> getStatus,
+            Func<Response<TResult>, TValue> parseResponse,
+            Func<Response<TResult>, RequestFailedException> getFailure,
+            TimeSpan? defaultPollingInterval = null)
+        {
+            _operationTypeName = operationTypeName;
+            _clientDiagnostics = clientDiagnostics;
+            _getResponseAsync = getResponseAsync;
+            _getResponse = getResponse;
+            _getStatus = getStatus;
+            _parseResponse = parseResponse;
+            _getFailure = getFailure;
+            _defaultPollingInterval = defaultPollingInterval ?? TimeSpan.FromSeconds(1);
+        }
+
+        public bool HasCompleted { get; private set; }
+
+        public bool HasValue { get; private set; }
+
+        public TValue Value
+        {
+            get
+            {
+                if (HasValue)
+                {
+                    return _value;
+                }
+                else if (_requestFailedException != null)
+                {
+                    throw _requestFailedException;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The operation has not completed yet.");
+                }
+            }
+        }
+
+        public Response GetRawResponse() => _rawResponse;
+
+        public async ValueTask<Response> UpdateStatusAsync(CancellationToken cancellationToken) =>
+            await UpdateStatusAsync(async: true, cancellationToken).ConfigureAwait(false);
+
+        public Response UpdateStatus(CancellationToken cancellationToken) =>
+            UpdateStatusAsync(async: false, cancellationToken).EnsureCompleted();
+
+        public async ValueTask<Response<TValue>> WaitForCompletionAsync(CancellationToken cancellationToken) =>
+            await WaitForCompletionAsync(_defaultPollingInterval, cancellationToken).ConfigureAwait(false);
+
+        public async ValueTask<Response<TValue>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                await UpdateStatusAsync(cancellationToken).ConfigureAwait(false);
+                if (HasCompleted)
+                {
+                    return Response.FromValue(Value, GetRawResponse());
+                }
+
+                await Task.Delay(pollingInterval, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
```

Comment:
Flow of `UpdateStatus` in multiple clients goes as follows:
- Check if already completed. If not, poll.
- If operation succeeded, parse response.
- If failed, throw (we could still throw with code 2xx depending on the content of the response).

Reply:
Added (a lot of) docstrings to the methods we expect developers to use. We still intend to add a usage guide in our docs in another PR.